logsapi/functionlogs_test.go

```diff
@@ -0,0 +1,54 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package logsapi
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/elastic/apm-aws-lambda/apmproxy"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestProcessFunctionLog(t *testing.T) {
+	metadataContainer := &apmproxy.MetadataContainer{
+		Metadata: []byte(`{"metadata":{"service":{"agent":{"name":"apm-lambda-extension","version":"1.1.0"},"framework":{"name":"AWS Lambda","version":""},"language":{"name":"python","version":"3.9.8"},"runtime":{"name":"","version":""},"node":{}},"user":{},"process":{"pid":0},"system":{"container":{"id":""},"kubernetes":{"node":{},"pod":{}}},"cloud":{"provider":"","instance":{},"machine":{},"account":{},"project":{},"service":{}}}}`),
+	}
+	event := LogEvent{
+		Time:         time.Date(2022, 11, 12, 0, 0, 0, 0, time.UTC),
+		Type:         FunctionLog,
+		StringRecord: "ERROR encountered. Stack trace:my-function (line 10)",
+	}
+	reqID := "8476a536-e9f4-11e8-9739-2dfe598c3fcd"
+	invokedFnArn := "arn:aws:lambda:us-east-2:123456789012:function:custom-runtime"
+	expectedData := fmt.Sprintf(
+		"%s\n{\"log\":{\"message\":\"%s\",\"@timestamp\":%d,\"faas\":{\"coldstart\":false,\"execution\":\"%s\",\"id\":\"%s\"}}}",
```

Comment:
[**For reviewers**] `coldstart` value here might not be correct and (CMIIW) I don't think there is a way for us to know about this by using `function` log event type. However, the default marshal logic always adds `coldstart` field.

I am thinking it might be better for us to have a custom unmarshal for `FAAS` here. Happy to discuss other thoughts on this.

Reply:
While testing the feature, I noticed that it is possible to get `function` log events while a different request ID is being processed. The payload of `function` log doesn't include any `requestID` for correlation with other events.

Based on current knowledge, I think we will have to drop `FAAS` fields from the log body completely.

Reply:
For posterity, discussed this with @axw and concluded with the following:

1. Add a custom Marshal logic for `faas`
2. Use `platform.start` and `platform.end` events to track request ID for function logs (assuming function logs will always be bounded between these two events)

Reply:
that sounds like a smart solution!

Reply:
This is fixed now.